internal/gstreamer-sink/gst.go

```diff
@@ -0,0 +1,66 @@
+// Package gst provides an easy API to create an appsrc pipeline
+package gst
+
+/*
+#cgo pkg-config: gstreamer-1.0 gstreamer-app-1.0
+
+#include "gst.h"
+
+*/
+import "C"
+import (
+	"unsafe"
+)
+
+// StartMainLoop starts GLib's main loop
+// It needs to be called from the process' main thread
+// Because many gstreamer plugins require access to the main thread
+// See: https://golang.org/pkg/runtime/#LockOSThread
+func StartMainLoop() {
+	C.gstreamer_receive_start_mainloop()
+}
+
+// Pipeline is a wrapper for a GStreamer Pipeline
+type Pipeline struct {
+	Pipeline *C.GstElement
+}
+
+// CreatePipeline creates a GStreamer Pipeline
+func CreatePipeline(codecName string) *Pipeline {
+	pipelineStr := "appsrc format=time is-live=true do-timestamp=true name=src ! application/x-rtp"
+	switch codecName {
+	case "vp8":
+		pipelineStr += ", encoding-name=VP8-DRAFT-IETF-01 ! rtpvp8depay ! decodebin ! autovideosink"
+	case "opus":
+		pipelineStr += ", payload=96, encoding-name=OPUS ! rtpopusdepay ! decodebin ! autoaudiosink"
+	case "vp9":
+		pipelineStr += " ! rtpvp9depay ! decodebin ! autovideosink"
+	case "h264":
+		pipelineStr += " ! rtph264depay ! decodebin ! autovideosink"
+	case "g722":
+		pipelineStr += " clock-rate=8000 ! rtpg722depay ! decodebin ! autoaudiosink"
+	default:
+		panic("Unhandled codec " + codecName)
+	}
+
```

Comment:
i think i do not understand, because i am very new to gst; is this pipeline what everyone will ultimately use?

1) when i read the `gst-send` example from pion, i thought that this was just a "demo" pipeline (and people would want to pass their own pipeline strings) - is there a way to use each video track after it has been sent to `autovideosink` here? does it belong in `pkg/` (because it is reusable) or does it belong in `examples/` (because it is demonstration)?

2) i tried to create this same example PR #20 (it should be called "stream-to-gst"), but i thought that most people would want to do both audio + video at the same time into the pipeline (my own goal is to send A + V to RTMP, ultimately) - so i try to write 2 tracks A + V to the pipeline, still WIP... for example is there any way we can use ^^ these 1-track-per-pipeline to write a single RTMP pipeline, or save a file to disk with audio + video at the same time?

Reply:
Im new to gst too.
Gstreamer works good in client side: pipeline+elements